.github/workflows/maven-release.yaml

```diff
@@ -0,0 +1,31 @@
+# This workflow will put the project in our staging repo
+name: Releasing Project to maven
+
+on:
+  release:
+    types: [ published ]
+
+jobs:
+  build:
+    if: endsWith(github.ref, "-lib")
+    env:
+      AWS_DEFAULT_REGION: us-east-1
+      AWS_REGION: us-east-1
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Java & publishing credentials
+      uses: actions/setup-java@v1
+      with:
+        java-version: 8
+        server-id: sonatype-nexus-staging # Value of the distributionManagement/repository/id field of the pom.xml
+        server-username: SONATYPE_USERNAME # env variable for username in deploy
+        server-password: SONATYPE_PASSWORD  # env variable for token in deploy
+        gpg-private-key: ${{ secrets.GPG_PRIVATE_KEY }} # Value of the GPG private key to import
+        gpg-passphrase: MAVEN_GPG_PASSPHRASE # env variable for GPG private key passphrase
+    - name: Deploy to sonatype staging repo
+      run: mvn deploy -Ppublishing
+      env:
+        SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
+        SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
+        MAVEN_GPG_PASSPHRASE: ${{ secrets.GPG_PASSPHRASE }}
```

Comment:
confused what's going on with [`secrets`](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)/[`env variables`](https://docs.github.com/en/free-pro-team@latest/actions/reference/environment-variables) here

looks similar to [`actions/setup-java` README](https://github.com/actions/setup-java#publishing-using-apache-maven) but don't understand why `env` variables are set in the step after they're used 🤔

Reply:
i think it's to limit the scope of when these secrets are on the build machine. the  setup-java step simply creates a configuration file which points to environment variables then the deploy step sets these when the configuration file is actually used (during the `mvn deploy` command)